esmvaltool/diag_scripts/weighting/climwip/core_functions.py

```diff
@@ -162,19 +196,52 @@ def calculate_weights(
     """
     if performance is not None:
```

Comment:
To me it is a bit clearer if you set the default values for numerator/denominator/not_nan first, and then change them depending if independence/performance is specified. You can also argue if independence / performance are mutually exclusive parameters whether they deserve to have their own functions to make this distinction clear.

```python
numerator = 1
not_nan = True
denominator = 1

if performance is not None:
        numerator = np.exp(-((performance / performance_sigma)**2))
        # nans in the performance vector indicate models to be excluded
        not_nan = np.isfinite(performance)

if independence is not None:
    # don't consider nan models for independence of other models!
    exp = np.exp(-((independence[:, not_nan] / independence_sigma)**2))
    # Note diagonal = exp(0) = 1, thus this is equal to 1 + sum(i!=j)
    denominator = exp.sum(axis=1)
```